File: Python/Binary Tree Level Order Traversal II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        queue = [(root, 1)]
        result_list = []
        for node, level in queue:
            if node:
                if len(result_list) < level:
                    result_list.insert(0, [])
                result_list[-level].append(node.val)
                if node.left:
                    queue.append((node.left, level + 1))
                if node.right:
                    queue.append((node.right, level + 1))
                
        
        return result_list
                
            
        

File: cpp/Binary Tree Level Order Traversal II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        queue<pair<TreeNode*, size_t >> q;
        q.push(pair<TreeNode*, size_t >(root, 1));
        vector<vector<int>> result_list;

        while (!q.empty()) {
            pair<TreeNode*, size_t > p = (pair<TreeNode *, size_t> &&) q.front();
            q.pop();
            if (p.first) {
                if (result_list.size() < p.second) {
                    vector<int> temp;
                    result_list.push_back(temp);
                }
                result_list[p.second - 1].push_back(p.first->val);
                if (p.first->left)
                    q.push(pair<TreeNode*, size_t >(p.first->left, p.second + 1));
                if (p.first->right)
                    q.push(pair<TreeNode*, size_t >(p.first->right, p.second + 1));
            }
        }
        reverse(result_list.begin(), result_list.end());
        return result_list;
    }
};
